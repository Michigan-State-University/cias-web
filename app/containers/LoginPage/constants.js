export const LOGIN_FORM_VIEW = 'LOGIN_FORM_VIEW';
export const CODE_VERIFICATION_VIEW = 'CODE_VERIFICATION_VIEW';
export const TERMS_NOT_ACCEPTED_VIEW = 'TERMS_NOT_ACCEPTED_VIEW';
