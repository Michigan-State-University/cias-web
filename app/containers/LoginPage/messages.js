/*
 * LoginPage Messages
 *
 * This contains all the text for the LoginPage container.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.containers.LoginPage';

export default defineMessages({
  pageTitle: {
    id: `${scope}.pageTitle`,
    defaultMessage: 'Login Page',
  },
  header: {
    id: `${scope}.header`,
    defaultMessage: 'Login',
  },
  loginButton: {
    id: `${scope}.loginButton`,
    defaultMessage: 'Log in',
  },
  emailPlaceholder: {
    id: `${scope}.emailPlaceholder`,
    defaultMessage: 'Email',
  },
  passwordPlaceholder: {
    id: `${scope}.passwordPlaceholder`,
    defaultMessage: 'Password',
  },
});
