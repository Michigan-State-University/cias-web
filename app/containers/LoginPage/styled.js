import styled from 'styled-components';

import Text from 'components/Text';

export const Or = styled(Text)`
  align-self: flex-end;
  line-height: 17px;
`;
