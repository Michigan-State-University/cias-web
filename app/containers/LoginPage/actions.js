/*
 *
 * LoginPage actions
 *
 */

import { LOGIN_REQUEST, LOGIN_SUCCESS, LOGIN_ERROR } from './constants';
import { actionBuilder } from '../../utils/actionBuilder';

const loginRequest = (username, password) =>
  actionBuilder(LOGIN_REQUEST, { username, password });

const loginSuccess = token => actionBuilder(LOGIN_SUCCESS, { token });

const loginError = error => actionBuilder(LOGIN_ERROR, { error });

export { loginRequest, loginSuccess, loginError };
