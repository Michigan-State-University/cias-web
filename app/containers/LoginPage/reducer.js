/*
 *
 * LoginPage reducer
 *
 */
import produce from 'immer';
import { REGISTER_SUCCESS } from 'containers/RegisterPage/constants';
import { LOGIN_REQUEST, LOGIN_SUCCESS, LOGIN_ERROR } from './constants';

export const initialState = {
  loading: false,
  error: '',
  newAccountPopup: false,
  formData: {
    email: '',
    password: '',
  },
};

/* eslint-disable default-case, no-param-reassign */
const loginPageReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOGIN_REQUEST:
        draft.formData = {
          email: action.payload.email,
          password: action.payload.password,
        };
        draft.error = '';
        draft.loading = true;
        break;
      case LOGIN_SUCCESS:
        draft.error = '';
        draft.loading = false;
        break;
      case LOGIN_ERROR:
        draft.loading = false;
        draft.error = action.payload.error;
        break;
      case REGISTER_SUCCESS:
        draft.newAccountPopup = true;
        break;
    }
  });

export default loginPageReducer;
