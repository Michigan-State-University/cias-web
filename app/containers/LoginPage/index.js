/**
 *
 * LoginPage
 *
 */

import React, { useCallback, useEffect, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useIntl } from 'react-intl';
import { parse } from 'query-string';
import has from 'lodash/has';
import { toast } from 'react-toastify';
import { useLocation } from 'react-router';
import { useInjectSaga } from 'redux-injectors';

import { Fill } from 'components/Fill';
import Column from 'components/Column';
import H1 from 'components/H1';
import { PublicLayout } from 'containers/PublicLayout';

import {
  loginRequest,
  makeSelectErrors,
  makeSelectLoaders,
  makeSelectLoginFormData,
  ACCOUNT_CONFIRMATION_ERROR,
  ACCOUNT_CONFIRMATION_SUCCESS,
  loginSaga,
  makeSelectVerificationNeeded,
  verificationCodeRequest,
  makeSelectVerificationSuccess,
} from 'global/reducers/auth';

import LoginForm from './components/LoginForm';
import CodeVerification from './components/CodeVerification';
import messages from './messages';
import { CODE_VERIFICATION_VIEW, LOGIN_FORM_VIEW } from './constants';

export const LoginPage = ({
  formData,
  onLogin,
  errors: { loginError, verificationCodeError },
  loaders: { loginLoading, verificationCodeLoading },
  verificationNeeded,
  verificationSuccess,
  verifyCode,
}) => {
  useInjectSaga({ key: 'loginPage', saga: loginSaga });
  const { search } = useLocation();
  const { formatMessage } = useIntl();

  const [currentView, setCurrentView] = useState(LOGIN_FORM_VIEW);
  const toLoginFormView = () => setCurrentView(LOGIN_FORM_VIEW);
  const toCodeVerificationView = () => setCurrentView(CODE_VERIFICATION_VIEW);

  useEffect(() => {
    if (verificationNeeded) toCodeVerificationView();
  }, [verificationNeeded]);

  useEffect(() => {
    if (verificationSuccess) {
      toLoginFormView();
      onLogin(formData.email, formData.password);
    }
  }, [verificationSuccess]);

  const queryObject = useMemo(
    () => parse(search, { parseBooleans: true }),
    [search],
  );

  useEffect(() => {
    if (has(queryObject, 'account_confirmation_success')) {
      if (queryObject.account_confirmation_success)
        toast.success(formatMessage(messages.accountConfirmation.success), {
          toastId: ACCOUNT_CONFIRMATION_SUCCESS,
        });
      else
        toast.error(formatMessage(messages.accountConfirmation.error), {
          toastId: ACCOUNT_CONFIRMATION_ERROR,
        });
    }
  }, [queryObject.account_confirmation_success]);

  const render = useCallback(() => {
    switch (currentView) {
      case CODE_VERIFICATION_VIEW:
        return (
          <CodeVerification
            goBack={toLoginFormView}
            isLoading={verificationCodeLoading}
            error={verificationCodeError}
            verifyCode={verifyCode}
          />
        );
      case LOGIN_FORM_VIEW:
      default:
        return (
          <LoginForm
            formData={formData}
            isLoading={loginLoading}
            onLogin={onLogin}
            error={loginError}
          />
        );
    }
  }, [
    currentView,
    loginLoading,
    verificationCodeLoading,
    loginError,
    verificationCodeError,
  ]);

  return (
    <>
      <Helmet>
        <title>{formatMessage(messages.pageTitle)}</title>
      </Helmet>
      <PublicLayout withMsuLogo>
        <Fill justify="center" align="center">
          <Column sm={10} md={8} lg={6} align="start">
            <H1 mb={40} fontSize={23}>
              {formatMessage(messages.header)}
            </H1>
            {render()}
          </Column>
        </Fill>
      </PublicLayout>
    </>
  );
};

LoginPage.propTypes = {
  formData: PropTypes.shape({
    email: PropTypes.string,
    password: PropTypes.string,
  }),
  onLogin: PropTypes.func,
  errors: PropTypes.object,
  loaders: PropTypes.object,
  verificationNeeded: PropTypes.bool,
  verificationSuccess: PropTypes.bool,
  verifyCode: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  formData: makeSelectLoginFormData(),
  loaders: makeSelectLoaders(),
  errors: makeSelectErrors(),
  verificationNeeded: makeSelectVerificationNeeded(),
  verificationSuccess: makeSelectVerificationSuccess(),
});

const mapDispatchToProps = {
  onLogin: loginRequest,
  verifyCode: verificationCodeRequest,
};

const withConnect = connect(mapStateToProps, mapDispatchToProps);

export default compose(withConnect)(LoginPage);
