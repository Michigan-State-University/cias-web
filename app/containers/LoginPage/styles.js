import styled from 'styled-components';
import { Centered } from '../../components/Centered';

const FullPageCentered = styled(Centered)`
  flex-grow: 1;
`;

export { FullPageCentered };
