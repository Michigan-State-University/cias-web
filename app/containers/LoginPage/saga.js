import { put, takeLatest } from 'redux-saga/effects';
import { push } from 'connected-react-router';
import axios from 'axios';
import { LOGIN_REQUEST } from './constants';
import { loginError, loginSuccess } from './actions';
import { logIn } from '../../global/reducers/auth/actions';

function* login({ payload: { email, password } }) {
  const requestURL = `auth/sign_in`;

  try {
    const {
      data: { data },
    } = yield axios.post(requestURL, {
      email,
      password,
    });

    yield put(
      logIn({
        firstName: data.first_name,
        lastName: data.last_name,
      }),
    );
    yield put(loginSuccess());
    yield put(push('/'));
  } catch (error) {
    const errorMessage = error.response.data.errors[0];
    yield put(loginError(errorMessage));
  }
}

export default function* loginPageSaga() {
  yield takeLatest(LOGIN_REQUEST, login);
}
