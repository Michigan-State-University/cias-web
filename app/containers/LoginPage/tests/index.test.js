/**
 *
 * Tests for LoginPage
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
import { Provider } from 'react-redux';

import LoginPage from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';
import configureStore from '../../../configureStore';

describe('<LoginPage />', () => {
  let store;
  let dispatch;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
    dispatch = jest.fn();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');

    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <LoginPage dispatch={dispatch} />
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <LoginPage dispatch={dispatch} />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
