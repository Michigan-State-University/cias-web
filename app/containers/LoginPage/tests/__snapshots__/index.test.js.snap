// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginPage /> Should render and match the snapshot 1`] = `
<div
  class="Fill-sc-1f4o7x0-0 dHXeMo"
>
  <div
    class="Column-sc-1n6divy-0 lxsJW"
  >
    <div
      class="Card-v952e1-0 kutakG"
      width="100%"
    >
      <div
        class="Column-sc-1n6divy-0 kCIvWB"
      >
        <p>
          Login
        </p>
        <input
          class="Input-sc-19txxuq-0 AcTzB"
          placeholder="Email"
          type="text"
          value=""
        />
        <input
          class="Input-sc-19txxuq-0 AcTzB"
          placeholder="Password"
          type="password"
          value=""
        />
        <button
          class="StyledButton-t20ntd-0 gNPbpR"
          color="primary"
          title="Log in"
          width="100%"
        >
          Log in
        </button>
      </div>
    </div>
  </div>
</div>
`;
